ooniprobe/Utility/ProxySettings.m

```diff
@@ -0,0 +1,78 @@
+#import "ProxySettings.h"
+
+@implementation ProxySettings
+
+- (id) init {
+    self = [super init];
+    if (self) {
+        NSString *protocol = [[NSUserDefaults standardUserDefaults] objectForKey:@"proxy_protocol"];
+        if ([protocol isEqualToString:[ProxySettings getProtocol:NONE]]) {
+            self.protocol = NONE;
+        } else if ([protocol isEqualToString:[ProxySettings getProtocol:PSIPHON]]) {
+            self.protocol = PSIPHON;
+        } else if ([protocol isEqualToString:[ProxySettings getProtocol:SOCKS5]]) {
+            self.protocol = SOCKS5;
+        } else {
+            // This is where we will extend the code to add support for
+            // more proxies, e.g., HTTP proxies.
+        }
+        self.hostname = [[NSUserDefaults standardUserDefaults] objectForKey:@"proxy_hostname"];
+        self.port = [[NSUserDefaults standardUserDefaults] objectForKey:@"proxy_port"];
+    }
+    return self;
+}
+
+- (void)saveProxy{
+    [[NSUserDefaults standardUserDefaults] setObject:[ProxySettings getProtocol:self.protocol] forKey:@"proxy_protocol"];
+    [[NSUserDefaults standardUserDefaults] setObject:self.hostname forKey:@"proxy_hostname"];
+    [[NSUserDefaults standardUserDefaults] setObject:self.port forKey:@"proxy_port"];
+    [[NSUserDefaults standardUserDefaults] synchronize];
+}
+
+- (BOOL)isCustom{
+    if (self.protocol == SOCKS5)
+        return true;
+    return false;
+}
+
+- (NSString*)getProxyString{
+    if (self.protocol == NONE)
+        return @"";
+    if (self.protocol == PSIPHON)
+        return @"psiphon://";
+    if (self.protocol == SOCKS5) {
+        // Alright, we now need to construct a new SOCKS5 URL. We are going to defer
+        // doing that to the Java standard library (er, the Android stdlib).
+        NSString *urlStr = [NSString stringWithFormat:@"socks5://%@:%@/", self.hostname, self.port];
+        if ([ProxySettings isIPv6:self.hostname]) {
+            urlStr = [NSString stringWithFormat:@"socks5://[%@]:%@/", self.hostname, self.port]; // IPv6 must be quoted in URLs
+        }
+        //URI url = new URI(urlStr);
+        return urlStr;
+    }
+    return @"";
+
+}
+
++ (BOOL)isIPv6:(NSString*)hostname{
+    return false;
```

Comment:
I am not very convinced by this implementation

Reply:
I still don't know how to implement this, should we use a regexp?